File: clients/ios/Classes/AskAI/AskAIViewModel.swift
//
//  AskAIViewModel.swift
//  NewsBlur
//
//  Created by Claude on 2024-12-06.
//  Copyright © 2024 NewsBlur. All rights reserved.
//

import Foundation
import SwiftUI
import Combine

@MainActor
class AskAIViewModel: ObservableObject {
    // MARK: - Published Properties

    @Published var conversation: AskAIConversation
    @Published var selectedModel: AskAIProvider = .opus
    @Published var selectedSummarizeType: AskAIQuestionType = .bullets
    @Published var customQuestion: String = ""
    @Published var isRecording: Bool = false
    @Published var isTranscribing: Bool = false
    @Published var showModelPicker: Bool = false
    @Published var hasAskedQuestion: Bool = false

    // MARK: - Private Properties

    private let appDelegate = NewsBlurAppDelegate.shared()!
    private var socketSubscribed = false
    private var timeoutTimer: Timer?
    private let timeoutDuration: TimeInterval = 15.0
    private let streamingTimeoutDuration: TimeInterval = 20.0

    // Voice recorder
    let voiceRecorder = VoiceRecorder()

    // Store story for re-initialization
    let story: [String: Any]

    // MARK: - Initialization

    init(story: [String: Any]) {
        self.story = story
        let storyHash = story["story_hash"] as? String ?? ""
        let storyTitle = ((story["story_title"] as? String ?? "") as NSString).decodingHTMLEntities() ?? ""

        self.conversation = AskAIConversation(storyHash: storyHash, storyTitle: storyTitle)

        // Load saved model preference
        if let savedModel = UserDefaults.standard.string(forKey: "askAIModel"),
           let model = AskAIProvider(rawValue: savedModel) {
            self.selectedModel = model
        }

        setupSocketHandlers()
        setupVoiceRecorderCallbacks()
    }

    deinit {
        // Note: deinit runs outside MainActor, so we need to unsubscribe synchronously
        // This is safe because socket unsubscription doesn't require MainActor
        let socket = NewsBlurSocketClient.shared
        socket.unsubscribe(event: "ask_ai:start")
        socket.unsubscribe(event: "ask_ai:chunk")
        socket.unsubscribe(event: "ask_ai:complete")
        socket.unsubscribe(event: "ask_ai:usage")
        socket.unsubscribe(event: "ask_ai:error")
    }

    // MARK: - Socket Setup

    private func setupSocketHandlers() {
        guard !socketSubscribed else { return }

        let socket = NewsBlurSocketClient.shared

        socket.subscribe(event: "ask_ai:start") { [weak self] data in
            Task { @MainActor in
                self?.handleStart(data)
            }
        }

        socket.subscribe(event: "ask_ai:chunk") { [weak self] data in
            Task { @MainActor in
                self?.handleChunk(data)
            }
        }

        socket.subscribe(event: "ask_ai:complete") { [weak self] data in
            Task { @MainActor in
                self?.handleComplete(data)
            }
        }

        socket.subscribe(event: "ask_ai:usage") { [weak self] data in
            Task { @MainActor in
                self?.handleUsage(data)
            }
        }

        socket.subscribe(event: "ask_ai:error") { [weak self] data in
            Task { @MainActor in
                self?.handleError(data)
            }
        }

        socketSubscribed = true
    }

    private func unsubscribeFromSocket() {
        guard socketSubscribed else { return }

        let socket = NewsBlurSocketClient.shared
        socket.unsubscribe(event: "ask_ai:start")
        socket.unsubscribe(event: "ask_ai:chunk")
        socket.unsubscribe(event: "ask_ai:complete")
        socket.unsubscribe(event: "ask_ai:usage")
        socket.unsubscribe(event: "ask_ai:error")

        socketSubscribed = false
    }

    // MARK: - Voice Recorder Setup

    private func setupVoiceRecorderCallbacks() {
        voiceRecorder.onTranscriptionComplete = { [weak self] text in
            Task { @MainActor in
                self?.customQuestion = text
                self?.isRecording = false
                self?.isTranscribing = false

                // Auto-submit if we have text
                if !text.isEmpty {
                    self?.sendQuestion(.custom)
                }
            }
        }

        voiceRecorder.onTranscriptionError = { [weak self] error in
            Task { @MainActor in
                self?.isRecording = false
                self?.isTranscribing = false
                self?.conversation.error = error
            }
        }

        voiceRecorder.onRecordingStateChange = { [weak self] isRecording in
            Task { @MainActor in
                self?.isRecording = isRecording
            }
        }
    }

    // MARK: - Public Methods

    func sendQuestion(_ type: AskAIQuestionType) {
        // Reset state
        conversation.error = nil
        conversation.usageMessage = nil
        conversation.responseText = ""
        conversation.isStreaming = true
        conversation.isComplete = false
        conversation.requestId = UUID().uuidString
        conversation.model = selectedModel
        hasAskedQuestion = true

        // Set question info
        if type == .custom {
            conversation.questionId = "custom"
            conversation.questionText = customQuestion
        } else {
            conversation.questionId = type.rawValue
            conversation.questionText = type.questionDescription
        }

        // Start timeout
        startTimeout()

        // Save model preference
        UserDefaults.standard.set(selectedModel.rawValue, forKey: "askAIModel")

        // Make API request
        sendQuestionRequest()
    }

    func sendFollowUp() {
        guard !customQuestion.isEmpty else { return }

        // Reset for new response (responseText is empty after handleComplete saves to completedBlocks)
        conversation.isStreaming = true
        conversation.isComplete = false
        conversation.requestId = UUID().uuidString
        conversation.questionText = customQuestion
        conversation.questionId = "custom"
        conversation.model = selectedModel  // Use the currently selected model
        customQuestion = ""

        // Save model preference
        UserDefaults.standard.set(selectedModel.rawValue, forKey: "askAIModel")

        startTimeout()
        sendQuestionRequest()
    }

    func reaskWithModel(_ model: AskAIProvider) {
        selectedModel = model
        conversation.model = model
        conversation.responseText = ""
        conversation.isStreaming = true
        conversation.isComplete = false
        conversation.requestId = UUID().uuidString

        // Save model preference
        UserDefaults.standard.set(model.rawValue, forKey: "askAIModel")

        startTimeout()
        sendQuestionRequest()
    }

    func cancelRequest() {
        timeoutTimer?.invalidate()
        conversation.isStreaming = false
    }

    func reset() {
        let storyHash = conversation.storyHash
        let storyTitle = conversation.storyTitle
        conversation = AskAIConversation(storyHash: storyHash, storyTitle: storyTitle)
        customQuestion = ""
        hasAskedQuestion = false
    }

    // MARK: - Voice Recording

    func toggleRecording() {
        if isRecording {
            stopRecording()
        } else {
            startRecording()
        }
    }

    func startRecording() {
        Task {
            let hasPermission = await voiceRecorder.requestPermissions()
            if hasPermission {
                voiceRecorder.startRecording()
                isRecording = true
            } else {
                conversation.error = "Microphone permission denied"
            }
        }
    }

    func stopRecording() {
        isTranscribing = true
        voiceRecorder.stopRecording()
    }

    // MARK: - Private Methods

    private func sendQuestionRequest() {
        let baseURL = appDelegate.url ?? "https://www.newsblur.com"
        guard let url = URL(string: "\(baseURL)/ask-ai/question") else {
            conversation.error = "Invalid URL"
            conversation.isStreaming = false
            return
        }

        NSLog("AskAI: Sending request to \(url)")
        NSLog("AskAI: Socket connected: \(NewsBlurSocketClient.shared.connected)")

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")

        // Build parameters
        var params: [String: String] = [
            "story_hash": conversation.storyHash,
            "question_id": conversation.questionId,
            "model": selectedModel.rawValue,
            "request_id": conversation.requestId
        ]

        if conversation.questionId == "custom" {
            params["custom_question"] = conversation.questionText
        }

        if !conversation.conversationHistory.isEmpty {
            let historyData = conversation.conversationHistory.map { ["role": $0.role, "content": $0.content] }
            if let jsonData = try? JSONSerialization.data(withJSONObject: historyData),
               let jsonString = String(data: jsonData, encoding: .utf8) {
                params["conversation_history"] = jsonString
            }
        }

        let bodyString = params.map { "\($0.key)=\($0.value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? "")" }.joined(separator: "&")
        request.httpBody = bodyString.data(using: .utf8)

        // Add cookies for authentication
        if let cookies = HTTPCookieStorage.shared.cookies(for: url) {
            let cookieHeaders = HTTPCookie.requestHeaderFields(with: cookies)
            for (key, value) in cookieHeaders {
                request.setValue(value, forHTTPHeaderField: key)
            }
        }

        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            Task { @MainActor in
                guard let self = self else { return }

                if let error = error {
                    NSLog("AskAI: Request error: \(error.localizedDescription)")
                    self.conversation.error = error.localizedDescription
                    self.conversation.isStreaming = false
                    return
                }

                if let httpResponse = response as? HTTPURLResponse {
                    NSLog("AskAI: Response status: \(httpResponse.statusCode)")
                }

                guard let data = data else {
                    NSLog("AskAI: No response data")
                    self.conversation.error = "No response data"
                    self.conversation.isStreaming = false
                    return
                }

                if let responseStr = String(data: data, encoding: .utf8) {
                    NSLog("AskAI: Response body: \(responseStr.prefix(500))")
                }

                do {
                    let response = try JSONDecoder().decode(AskAIQuestionResponse.self, from: data)
                    NSLog("AskAI: Parsed response code: \(response.code)")

                    if response.code != 1 {
                        self.conversation.error = response.message ?? "Request failed"
                        self.conversation.isStreaming = false
                    }
                    // Success - wait for socket events
                } catch {
                    NSLog("AskAI: Parse error: \(error)")
                    // Try to parse error message
                    if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                       let message = json["message"] as? String {
                        self.conversation.error = message
                    } else {
                        self.conversation.error = "Failed to parse response"
                    }
                    self.conversation.isStreaming = false
                }
            }
        }.resume()
    }

    // MARK: - Socket Event Handlers

    private func handleStart(_ data: Any) {
        NSLog("AskAI: Received start event: \(data)")
        guard let dict = data as? [String: Any],
              let storyHash = dict["story_hash"] as? String,
              let requestId = dict["request_id"] as? String,
              storyHash == conversation.storyHash,
              requestId == conversation.requestId else {
            NSLog("AskAI: Start event ignored - mismatched hash/requestId")
            return
        }

        NSLog("AskAI: Start event matched, beginning streaming")
        // Reset timeout for streaming
        startStreamingTimeout()
        conversation.error = nil
    }

    private func handleChunk(_ data: Any) {
        guard let dict = data as? [String: Any],
              let storyHash = dict["story_hash"] as? String,
              let requestId = dict["request_id"] as? String,
              let chunk = dict["chunk"] as? String,
              storyHash == conversation.storyHash,
              requestId == conversation.requestId else {
            return
        }

        NSLog("AskAI: Received chunk: \(chunk.prefix(50))...")
        // Clear any timeout error since content is arriving
        if conversation.error != nil {
            conversation.error = nil
            conversation.isStreaming = true
        }

        // Append chunk to response
        conversation.responseText += chunk

        // Reset streaming timeout
        startStreamingTimeout()
    }

    private func handleComplete(_ data: Any) {
        NSLog("AskAI: Received complete event")
        guard let dict = data as? [String: Any],
              let storyHash = dict["story_hash"] as? String,
              let requestId = dict["request_id"] as? String,
              storyHash == conversation.storyHash,
              requestId == conversation.requestId else {
            return
        }

        NSLog("AskAI: Stream complete")
        timeoutTimer?.invalidate()

        let responseText = conversation.responseText
        let questionText = conversation.questionText

        // Save completed response to history
        let isFollowUp = !conversation.completedBlocks.isEmpty
        let block = AskAIResponseBlock(
            questionText: conversation.questionText,
            model: conversation.model,
            responseText: conversation.responseText,
            isFollowUp: isFollowUp
        )
        conversation.completedBlocks.append(block)

        // Preserve conversation context for follow-ups
        if !questionText.isEmpty {
            conversation.conversationHistory.append(
                AskAIMessage(role: "user", content: questionText)
            )
        }
        if !responseText.isEmpty {
            conversation.conversationHistory.append(
                AskAIMessage(role: "assistant", content: responseText)
            )
        }

        // Clear current response (it's now in completedBlocks)
        conversation.responseText = ""
        conversation.isStreaming = false
        conversation.isComplete = true
    }

    private func handleUsage(_ data: Any) {
        guard let dict = data as? [String: Any],
              let storyHash = dict["story_hash"] as? String,
              let requestId = dict["request_id"] as? String,
              let message = dict["message"] as? String,
              storyHash == conversation.storyHash,
              requestId == conversation.requestId else {
            return
        }

        conversation.usageMessage = message
    }

    private func handleError(_ data: Any) {
        guard let dict = data as? [String: Any],
              let storyHash = dict["story_hash"] as? String,
              let requestId = dict["request_id"] as? String,
              let error = dict["error"] as? String,
              storyHash == conversation.storyHash,
              requestId == conversation.requestId else {
            return
        }

        timeoutTimer?.invalidate()
        conversation.error = error
        conversation.isStreaming = false
    }

    // MARK: - Timeout Handling

    private func startTimeout() {
        timeoutTimer?.invalidate()
        timeoutTimer = Timer.scheduledTimer(withTimeInterval: timeoutDuration, repeats: false) { [weak self] _ in
            Task { @MainActor in
                guard let self = self, self.conversation.isStreaming, self.conversation.responseText.isEmpty else { return }
                self.conversation.error = "Request timed out"
                self.conversation.isStreaming = false
            }
        }
    }

    private func startStreamingTimeout() {
        timeoutTimer?.invalidate()
        timeoutTimer = Timer.scheduledTimer(withTimeInterval: streamingTimeoutDuration, repeats: false) { [weak self] _ in
            Task { @MainActor in
                guard let self = self, self.conversation.isStreaming else { return }

                // If we have received content, treat as successful even without complete event
                if !self.conversation.responseText.isEmpty {
                    NSLog("AskAI: Stream timeout but content received, completing gracefully")
                    let responseText = self.conversation.responseText
                    let questionText = self.conversation.questionText
                    // Save completed response to history
                    let isFollowUp = !self.conversation.completedBlocks.isEmpty
                    let block = AskAIResponseBlock(
                        questionText: self.conversation.questionText,
                        model: self.conversation.model,
                        responseText: self.conversation.responseText,
                        isFollowUp: isFollowUp
                    )
                    self.conversation.completedBlocks.append(block)
                    if !questionText.isEmpty {
                        self.conversation.conversationHistory.append(
                            AskAIMessage(role: "user", content: questionText)
                        )
                    }
                    if !responseText.isEmpty {
                        self.conversation.conversationHistory.append(
                            AskAIMessage(role: "assistant", content: responseText)
                        )
                    }
                    self.conversation.responseText = ""
                    self.conversation.isStreaming = false
                    self.conversation.isComplete = true
                } else {
                    self.conversation.error = "Stream interrupted"
                    self.conversation.isStreaming = false
                }
            }
        }
    }
}

File: clients/android/NewsBlur/app/src/main/java/com/newsblur/viewModel/AskAiViewModel.kt
package com.newsblur.viewModel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.newsblur.askai.AskAiMessage
import com.newsblur.askai.AskAiProvider
import com.newsblur.askai.AskAiQuestionType
import com.newsblur.askai.AskAiResponseBlock
import com.newsblur.askai.AskAiStory
import com.newsblur.askai.AskAiUiState
import com.newsblur.network.AskAiApi
import com.newsblur.network.AskAiQuestionRequest
import com.newsblur.network.NewsBlurSocketClient
import com.newsblur.preference.PrefsRepo
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.util.UUID
import javax.inject.Inject

@HiltViewModel
class AskAiViewModel
    @Inject
    constructor(
        private val askAiApi: AskAiApi,
        private val prefsRepo: PrefsRepo,
        private val socketClient: NewsBlurSocketClient,
    ) : ViewModel() {
        private val _uiState = MutableStateFlow(AskAiUiState())
        val uiState: StateFlow<AskAiUiState> = _uiState.asStateFlow()

        private val conversationHistory = mutableListOf<AskAiMessage>()
        private var pendingHistoryQuestion: AskAiMessage? = null
        private var socketSubscribed = false
        private var timeoutJob: Job? = null
        private var streamingTimeoutJob: Job? = null

        fun initialize(
            storyHash: String,
            storyTitle: String,
        ) {
            val currentStory = _uiState.value.story
            if (currentStory?.storyHash == storyHash) return

            conversationHistory.clear()
            pendingHistoryQuestion = null
            _uiState.value =
                AskAiUiState(
                    story = AskAiStory(storyHash = storyHash, storyTitle = storyTitle),
                    selectedModel = AskAiProvider.fromRawValue(prefsRepo.getAskAiModel()),
                    showAskAi = prefsRepo.isShowAskAi(),
                    isArchiveTier = prefsRepo.getIsArchive() || prefsRepo.getIsPro(),
                )

            setupSocketHandlers()
            prefsRepo.getUserName()?.let { socketClient.connect(username = it) }
        }

        fun updateCustomQuestion(question: String) {
            _uiState.update { it.copy(customQuestion = question) }
        }

        fun selectModel(model: AskAiProvider) {
            prefsRepo.setAskAiModel(model.rawValue)
            _uiState.update { it.copy(selectedModel = model) }
        }

        fun sendQuestion(type: AskAiQuestionType) {
            val customQuestion = _uiState.value.customQuestion.trim()
            val questionText =
                if (type == AskAiQuestionType.CUSTOM) {
                    customQuestion
                } else {
                    type.questionDescription
                }
            sendQuestion(
                questionId = type.rawValue,
                questionText = questionText,
                clearCustomQuestion = type == AskAiQuestionType.CUSTOM,
            )
        }

        fun sendFollowUp() {
            sendQuestion(AskAiQuestionType.CUSTOM)
        }

        fun reaskWithModel(model: AskAiProvider) {
            selectModel(model)
            val state = _uiState.value
            conversationHistory.clear()
            pendingHistoryQuestion = null
            sendQuestion(
                questionId = state.currentQuestionId,
                questionText = state.currentQuestionText,
                clearCustomQuestion = false,
            )
        }

        fun cancelRequest() {
            timeoutJob?.cancel()
            streamingTimeoutJob?.cancel()
            pendingHistoryQuestion = null
            _uiState.update { it.copy(isStreaming = false) }
        }

        fun setRecording(isRecording: Boolean) {
            _uiState.update {
                it.copy(
                    isRecording = isRecording,
                    isTranscribing = if (isRecording) false else it.isTranscribing,
                    errorMessage = if (isRecording) null else it.errorMessage,
                )
            }
        }

        fun setVoiceError(message: String) {
            _uiState.update {
                it.copy(
                    isRecording = false,
                    isTranscribing = false,
                    errorMessage = message,
                )
            }
        }

        fun transcribeAudio(file: File) {
            _uiState.update {
                it.copy(
                    isRecording = false,
                    isTranscribing = true,
                    errorMessage = null,
                )
            }

            viewModelScope.launch {
                val response =
                    withContext(Dispatchers.IO) {
                        askAiApi.transcribeAudio(file)
                    }
                file.delete()

                if (response.code == 1 && !response.text.isNullOrBlank()) {
                    _uiState.update {
                        it.copy(
                            customQuestion = response.text,
                            isTranscribing = false,
                        )
                    }
                    sendQuestion(AskAiQuestionType.CUSTOM)
                } else {
                    _uiState.update {
                        it.copy(
                            isTranscribing = false,
                            errorMessage = response.message ?: "Transcription failed",
                        )
                    }
                }
            }
        }

        override fun onCleared() {
            timeoutJob?.cancel()
            streamingTimeoutJob?.cancel()
            pendingHistoryQuestion = null
            socketClient.unsubscribe(EVENT_START)
            socketClient.unsubscribe(EVENT_CHUNK)
            socketClient.unsubscribe(EVENT_COMPLETE)
            socketClient.unsubscribe(EVENT_USAGE)
            socketClient.unsubscribe(EVENT_ERROR)
            super.onCleared()
        }

        private fun setupSocketHandlers() {
            if (socketSubscribed) return

            socketClient.subscribe(EVENT_START, ::handleStart)
            socketClient.subscribe(EVENT_CHUNK, ::handleChunk)
            socketClient.subscribe(EVENT_COMPLETE, ::handleComplete)
            socketClient.subscribe(EVENT_USAGE, ::handleUsage)
            socketClient.subscribe(EVENT_ERROR, ::handleError)
            socketSubscribed = true
        }

        private fun handleStart(data: Any?) {
            val payload = data.asPayload() ?: return
            if (!matchesCurrentRequest(payload)) return

            _uiState.update { it.copy(errorMessage = null) }
            startStreamingTimeout()
        }

        private fun handleChunk(data: Any?) {
            val payload = data.asPayload() ?: return
            if (!matchesCurrentRequest(payload)) return

            val chunk = payload["chunk"] as? String ?: return
            _uiState.update {
                it.copy(
                    currentResponseText = it.currentResponseText + chunk,
                    errorMessage = null,
                    isStreaming = true,
                )
            }
            startStreamingTimeout()
        }

        private fun handleComplete(data: Any?) {
            val payload = data.asPayload() ?: return
            if (!matchesCurrentRequest(payload)) return
            completeCurrentResponse()
        }

        private fun handleUsage(data: Any?) {
            val payload = data.asPayload() ?: return
            if (!matchesCurrentRequest(payload)) return

            _uiState.update { it.copy(usageMessage = payload["message"] as? String) }
        }

        private fun handleError(data: Any?) {
            val payload = data.asPayload() ?: return
            if (!matchesCurrentRequest(payload)) return

            timeoutJob?.cancel()
            streamingTimeoutJob?.cancel()
            pendingHistoryQuestion = null
            _uiState.update {
                it.copy(
                    isStreaming = false,
                    errorMessage = payload["error"] as? String ?: "Request failed",
                )
            }
        }

        private fun completeCurrentResponse() {
            timeoutJob?.cancel()
            streamingTimeoutJob?.cancel()

            val state = _uiState.value
            if (state.currentQuestionText.isBlank() && state.currentResponseText.isBlank()) return

            val block =
                AskAiResponseBlock(
                    questionText = state.currentQuestionText,
                    model = state.selectedModel,
                    responseText = state.currentResponseText,
                    isFollowUp = state.completedBlocks.isNotEmpty(),
                )

            pendingHistoryQuestion?.let { conversationHistory += it }
            if (state.currentResponseText.isNotBlank()) {
                conversationHistory += AskAiMessage(role = "assistant", content = state.currentResponseText)
            }
            pendingHistoryQuestion = null

            _uiState.update {
                it.copy(
                    completedBlocks = it.completedBlocks + block,
                    currentResponseText = "",
                    isStreaming = false,
                    isComplete = true,
                )
            }
        }

        private fun startTimeout() {
            timeoutJob?.cancel()
            timeoutJob =
                viewModelScope.launch {
                    kotlinx.coroutines.delay(REQUEST_TIMEOUT_MS)
                    val state = _uiState.value
                    if (state.isStreaming && state.currentResponseText.isBlank()) {
                        pendingHistoryQuestion = null
                        _uiState.update {
                            it.copy(
                                isStreaming = false,
                                errorMessage = "Request timed out",
                            )
                        }
                    }
                }
        }

        private fun startStreamingTimeout() {
            streamingTimeoutJob?.cancel()
            streamingTimeoutJob =
                viewModelScope.launch {
                    kotlinx.coroutines.delay(STREAMING_TIMEOUT_MS)
                    val state = _uiState.value
                    if (!state.isStreaming) return@launch

                    if (state.currentResponseText.isNotBlank()) {
                        completeCurrentResponse()
                    } else {
                        pendingHistoryQuestion = null
                        _uiState.update {
                            it.copy(
                                isStreaming = false,
                                errorMessage = "Stream interrupted",
                            )
                        }
                    }
                }
        }

        private fun matchesCurrentRequest(payload: Map<String, Any?>): Boolean {
            val state = _uiState.value
            val storyHash = payload["story_hash"] as? String
            val requestId = payload["request_id"] as? String
            return storyHash == state.story?.storyHash && requestId == state.currentRequestId
        }

        private fun sendQuestion(
            questionId: String,
            questionText: String,
            clearCustomQuestion: Boolean,
        ) {
            val story = _uiState.value.story ?: return
            if (questionId.isBlank() || questionText.isBlank()) return

            val requestId = UUID.randomUUID().toString()
            val selectedModel = _uiState.value.selectedModel
            val requestConversationHistory =
                if (conversationHistory.isNotEmpty()) {
                    val userMessage = AskAiMessage(role = "user", content = questionText)
                    pendingHistoryQuestion = userMessage
                    conversationHistory.toList() + userMessage
                } else {
                    pendingHistoryQuestion = null
                    emptyList()
                }

            _uiState.update {
                it.copy(
                    currentQuestionId = questionId,
                    currentQuestionText = questionText,
                    currentRequestId = requestId,
                    currentResponseText = "",
                    isStreaming = true,
                    isComplete = false,
                    hasAskedQuestion = true,
                    errorMessage = null,
                    usageMessage = null,
                    customQuestion = if (clearCustomQuestion) "" else it.customQuestion,
                )
            }
            startTimeout()

            viewModelScope.launch {
                val response =
                    withContext(Dispatchers.IO) {
                        askAiApi.sendQuestion(
                            AskAiQuestionRequest(
                                storyHash = story.storyHash,
                                questionId = questionId,
                                requestId = requestId,
                                model = selectedModel.rawValue,
                                customQuestion = questionText.takeIf { questionId == AskAiQuestionType.CUSTOM.rawValue },
                                conversationHistory = requestConversationHistory,
                            ),
                        )
                    }

                if (response.code != 1) {
                    timeoutJob?.cancel()
                    pendingHistoryQuestion = null
                    _uiState.update {
                        it.copy(
                            isStreaming = false,
                            errorMessage = response.message ?: "Request failed",
                        )
                    }
                }
            }
        }

        @Suppress("UNCHECKED_CAST")
        private fun Any?.asPayload(): Map<String, Any?>? = this as? Map<String, Any?>

        companion object {
            private const val EVENT_START = "ask_ai:start"
            private const val EVENT_CHUNK = "ask_ai:chunk"
            private const val EVENT_COMPLETE = "ask_ai:complete"
            private const val EVENT_USAGE = "ask_ai:usage"
            private const val EVENT_ERROR = "ask_ai:error"

            private const val REQUEST_TIMEOUT_MS = 15_000L
            private const val STREAMING_TIMEOUT_MS = 20_000L
        }
    }
